### PR TITLE
romio/daos: fix a double free issue when path has "." in it.

### PR DESCRIPTION
**Pull request by mchaarawi**
## Pull Request Description
romio/daos: fix a double free issue when path has "." in it.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/warnings
test:mpich/ch3/most
test:mpich/ch4/most

File: src/mpi/romio/adio/ad_daos/ad_daos_open.c [hunk 1]
```diff
@@ -45,8 +45,14 @@ static int parse_filename(const char *path, char **_obj_name, char **_cont_name)
                 rc = ENOMEM;
                 goto out;
             }
+            *_cont_name = cont_name;
+        } else {
+            *_cont_name = ADIOI_Strdup(cont_name);
+            if (*_cont_name == NULL) {
+                rc = ENOMEM;
+                goto out;
+            }
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_daos/ad_daos_open.c:54]: You just need to change the condition at line 32 into `if (strcmp(cont_name, ".") == 0)`, right?

Comment (mchaarawi) [reply src/mpi/romio/adio/ad_daos/ad_daos_open.c:54]: yes you are right.
but this case requires some more handling under this if condition to support properly which we don't yet, which is why i had it that way to just avoid the double free crash for now. anyway i will update this now.

File: src/mpi/romio/adio/ad_daos/ad_daos_open.c [hunk 2]
```diff
@@ -45,6 +45,21 @@ static int parse_filename(const char *path, char **_obj_name, char **_cont_name)
...
+        } else {
+            char *new_dir = ADIOI_Calloc(strlen(cwd) + strlen(cont_name) + 1, sizeof(char));
+            if (new_dir == NULL) {
+                rc = -ENOMEM;
+                goto out;
+            }
+
+            strcpy(new_dir, cwd);
+            if (cont_name[0] == '.') {
+                strcat(new_dir, &cont_name[1]);
```
Comment (hzhou) [inline src/mpi/romio/adio/ad_daos/ad_daos_open.c:57]: What about `../a_file_name`?

**Review by hzhou — COMMENTED** [review]
A comment

File: src/mpi/romio/adio/ad_daos/ad_daos_open.c (hunk 2 above)
Comment (mchaarawi) [reply src/mpi/romio/adio/ad_daos/ad_daos_open.c:57]: it's probably easier if i just use realpath(). let me do some testing and ask for reviews internally so i don't waste more of your time. thanks for your reviews.

**Comment by mchaarawi** [comment]
> A comment

should be good now

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Review by hzhou — APPROVED** [review]
Looks good to me. I'll merge it once the tests are cleared.

**Review by hzhou — APPROVED** [review]
@mchaarawi Could you rebase the PR?

**Comment by mchaarawi** [comment]
> @mchaarawi Could you rebase the PR?

done

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Pull request merged**